### ISSUE TITLE
Suggestion: All services should be registered in the main application container w/ no programmatic fallback

### ISSUE THREAD

**Issue by jeremydmiller** [aspnet/DependencyInjection#238]
tl;dr: I think the fallback mechanism for default services in the IoC abstraction is inefficient, potentially confusing, and makes IoC diagnostics less useful.

The main `GetService(type)` method of one of the IoC container abstractions today looks like this:

```
        public object GetService(Type serviceType)
        {
            var instance = _container.TryGetInstance(serviceType);
            if (instance != null)
            {
                return instance;
            }
            if (serviceType.IsGenericType && serviceType.GetGenericTypeDefinition()== typeof(IEnumerable<>))
            {
                return _container.GetInstance(serviceType);
            }

            return null;
        }
```

The code above is part of an effort to build a StructureMap plugin to MVC6. From my perspective, the code above is unnecessarily inefficient because of the call to `TryGetInstance` plus the if/then logic executed on every single call. From a purely IoC container perspective, I'd strongly prefer to see the method above for the StructureMap implementation be:

```
        public object GetService(Type serviceType)
        {
            // StructureMap can take care of the IEnumerable and generic types
            // without any helper logic
            return _container.GetInstance(serviceType);
        }
```

**Why?**

* The call to TryGetInstance() is inefficient because it has to scan the Container internal model to see if it could satisfy the request instead of just going off of a memoized "build plan" for that service type. I tell users that it's effectively an anti-pattern that's only there for frameworks like MVC.

* Using TryGetInstance() sidesteps some diagnostics built into StructureMap to explain to you why you're not getting a default type for the requested type. Using TryGetInstance with the fallback may very well be hiding information from users

* Using the programmatic fallback if the service resolution returns a null will make the behavior of the framework less obvious and will, IMO, cause support problems later. If all the services are required to be in the main application container, you can lean on the comprehensive diagnostics that most of the popular IoC containers provide. In later versions of FubuMVC, we were able to provide a diagnostics page showing all the services registered in the main app container so you always knew if a running application was using your overrides for built in services or the defaults from the framework itself. I think you're going to want this same functionality for MVC6 when folks start coming to you with problems when they try to override parts of the framework.

* For the sake of performance and understandability, we took the approach in FubuMVC to eliminate as much branching if/then logic at runtime as possible. In this case, you do know what the service types are upfront at configuration time, so you could simplify the interactions with the container by simply registering the defaults into the main application container. 

**How**

I'm guessing the root issue is that you have certain services that are required by the framework itself like `Microsoft.AspNet.DataProtection.XmlEncryption.IXmlEncryptor` that can be overridden by users, but uses default implementations if the user has not supplied an override. Having the fallback from the IServiceProvider returning null is one way to do this, but we were able to accomplish the same thing in FubuMVC by adding a new service registration concept called "SetServiceIfNone<T>" as shown here: https://github.com/DarthFubuMVC/fubumvc/blob/master/src/FubuMVC.Core/CoreServiceRegistry.cs. When a FubuMVC application is bootstrapped, the default service registrations are only applied if the application provides no overrides.

You already have a service registration model in the new framework, so what you could do at configuration time is to take the service registrations supplied by the user and add the default registrations for all the services required by the FX that were not supplied by the user to backfill the main container with whatever is missing.

**Comment by davidfowl** [comment]
I think there are 2 different sets of issues being called out here:

> tl;dr: I think the fallback mechanism for default services in the IoC abstraction is inefficient, potentially confusing, and makes IoC diagnostics less useful.

I agree and we're aware of this current limitation. We might be able to change GetService and GetRequiredService to GetService (bare metal behaviors of the DI container) and TryGetService (an extension method that checks for `IEnumerable<T>` so things can choose to fallback). Another thing that was discussed a while back was supporting default parameters in ctors:

```C#
public class MyClass()
{
    public MyClass(IFoo foo = null)
    {
    }
}
```

This would indicate that IFoo was optional. We weren't sure we if could rely on this in our frameworks because of the number of DI containers that potentially don't support that. 

That said, I'd need to check, but i'm pretty sure we don't do this many places in the stack. DataProtection is one of the worst offenders because it works outside of a DI context.

> I'm guessing the root issue is that you have certain services that are required by the framework itself like Microsoft.AspNet.DataProtection.XmlEncryption.IXmlEncryptor that can be overridden by users, but uses default implementations if the user has not supplied an override. Having the fallback from the IServiceProvider returning null is one way to do this, but we were able to accomplish the same thing in FubuMVC by adding a new service registration concept called "SetServiceIfNone" as shown here: https://github.com/DarthFubuMVC/fubumvc/blob/master/src/FubuMVC.Core/CoreServiceRegistry.cs. When a FubuMVC application is bootstrapped, the default service registrations are only applied if the application provides no overrides.

This is much easier said than done. FubuMVC was a single, self contained framework so it was easier to enforce that the application container needed to be setup fully by the end user in order to get working behavior. ASP.NET 5 is a bunch of different packages that can be used together with or without DI (in some cases). We also don't have a single component that understands all of the dependencies of each of the frameworks. The application is the only place you get a chance to compose disparate things.

The big question is, where do you set those defaults?  Fubu's CoreServiceRegistry is like a base class that understands the entire universe of services. We have no equivalent. The user has to take action here.

/cc @lodejard

**Comment by CoreyKaylor** [comment]
CoreServiceRegistry was one place that default services were registered. The external or optional components also had an ability to register their own default components. In other words it doesn't have to  imply a direct coupling as you have indicated.

**Comment by CoreyKaylor** [comment]
I think the "with or without DI" is the challenge here. I think littering the runtime with null checks for things that expected for a component to work in my opinion isn't a good way to deal with that.

Maybe there are concrete examples seen as problematic that can be talked through?

**Comment by davidfowl** [comment]
> CoreServiceRegistry was one place that default services were registered. The external or optional components also had an ability to register their own default components. In other words it doesn't have to imply a direct coupling as you have indicated.

How? We don't do any assembly scanning and don't intend to because we're sensitive about startup performance.

> I think the "with or without DI" is the challenge here. I think littering the runtime with null checks for things that expected for a component to work in my opinion isn't a good way to deal with that.

I agree, not many places do this but data protection is one of them. One of the ideas is that you can use some of these libraries in a console application (some are portable).

**Comment by CoreyKaylor** [comment]
Although Fubu did do assembly scanning in some cases to make it easier to "drop-in" new features. It wasn't necessary and could have been done in the equivalent way you might do `app.UseWelcomePage()` today. The programatic way it was done in Fubu was to call `registry.Import<WelcomePage>()`. The import call itself had the same control over service registry etc. that the main application had.

I'll see if I can set aside some time this weekend to take a look at the DataProtection libs. We're in the middle of selling our house though so my time seems to be more limited at the moment...

**Comment by lodejard** [comment]
Doing issue bookkeeping. @jeremydmiller @CoreyKaylor - is this issue still an active concern?

**Comment by jeremydmiller** [comment]
@lodejard Well, yeah, it is, but this was just a suggestion. From the StructureMap perspective, I'm really not wild about how vNext requires the `TryGetInstance` mechanism because it defeats some diagnostics and it's a touch less efficient. I think you could have gained some capabilities and had less confusion about runtime behavior down the line by having all the services inside the IoC container instead of depending on the fall through mechanisms.

**Comment by lodejard** [comment]
Well, I guess I'm going to close this because it's not really clear to me what change this is calling for. I believe it's agreed that using constructor injection and required services w/out a programatic fallback is preferred - but this issue itself isn't really actionable.

If folks see places in the code where this happens and should be changed please file issues there. And of course PRs are always welcome!
